File: frontend/src/services/attendanceService.js
import api from './api';

export const attendanceService = {
  // Mark attendance (Teacher/CA)
  markAttendance: async (attendanceData) => {
    return await api.post('/attendance/mark', attendanceData);
  },

  // Update attendance record (Teacher/CA)
  updateAttendance: async (attendanceId, status) => {
    return await api.put(`/attendance/${attendanceId}?status=${status}`);
  },

  // Submit attendance for a class (Teacher/CA)
  submitAttendance: async (classId, date) => {
    return await api.post(`/attendance/submit?classId=${classId}&date=${date}`);
  },

  // Get attendance by class and date
  getAttendanceByClassAndDate: async (classId, date) => {
    return await api.get(`/attendance/class/${classId}/date/${date}`);
  },

  // Get student attendance history
  getStudentAttendance: async (studentId, page = 0, size = 10) => {
    return await api.get(`/attendance/student/${studentId}?page=${page}&size=${size}`);
  },

  // Get attendance statistics for a class
  getAttendanceStatistics: async (classId, startDate, endDate) => {
    return await api.get(`/attendance/statistics/class/${classId}?startDate=${startDate}&endDate=${endDate}`);
  },

  // Get student attendance percentage
  getStudentAttendancePercentage: async (studentId, subjectId = null) => {
    let url = `/attendance/student/${studentId}/percentage`;
    if (subjectId) {
      url += `?subjectId=${subjectId}`;
    }
    return await api.get(url);
  },

  // Generate QR code for attendance (Teacher/CA)
  generateQRCode: async (classId, date) => {
    return await api.post(`/attendance/qr/generate?classId=${classId}&date=${date}`);
  },

  // Scan QR code for attendance (Student)
  scanQRCode: async (qrCode) => {
    return await api.post('/attendance/qr/scan', { qrCode });
  },

  // Export attendance report
  exportAttendanceReport: async (classId, startDate, endDate) => {
    return await api.get(`/attendance/export?classId=${classId}&startDate=${startDate}&endDate=${endDate}`, {
      responseType: 'blob',
    });
  },
};

File: backend/src/main/java/com/university/sms/service/AttendanceService.java
package com.university.sms.service;

import com.university.sms.dto.request.AttendanceRequest;
import com.university.sms.dto.response.AttendanceResponse;
import com.university.sms.dto.response.AttendanceStatisticsResponse;
import com.university.sms.dto.response.StudentAttendanceDashboardResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;

public interface AttendanceService {
    List<AttendanceResponse> markAttendance(AttendanceRequest request, Long teacherId);
    AttendanceResponse updateAttendance(Long attendanceId, String status, Long teacherId);
    void submitAttendance(Long classId, LocalDate date, Long teacherId);
    List<AttendanceResponse> getAttendanceByClassAndDate(Long classId, LocalDate date);
    Page<AttendanceResponse> getStudentAttendance(Long studentId, Pageable pageable);
    List<AttendanceResponse> getStudentAttendanceHistory(Long studentId);
    AttendanceStatisticsResponse getAttendanceStatistics(Long classId, LocalDate startDate, LocalDate endDate);
    Double getStudentAttendancePercentage(Long studentId, Long subjectId);
    String generateQrCodeForAttendance(Long classId, LocalDate date, Long teacherId);
    void processQrAttendance(String qrCode, Long studentId);
    void checkAndSendLowAttendanceAlerts();
    StudentAttendanceDashboardResponse getStudentAttendanceDashboard(Long studentId);
}